File: cppsrc/linear/ExpressionLowering.cpp
#include "ExpressionLowering.h"

#include <string>
#include <utility>

namespace linear {
    using ast::ExprKind;
    using ast::ExprPtr;
    using ast::SourcePos;
    using ast::Stmt;
    using ast::StmtKind;
    using ast::StmtPtr;

    namespace {
        // A literal, a name, a qualified name (`Res<T>`), or a lambda. A lambda's
        // own body is a separate body: the same passes lower it when it is emitted.
        bool isSimple(const ExprPtr &e) {
            switch (e->kind) {
                case ExprKind::IntLit:
                case ExprKind::FloatLit:
                case ExprKind::StrLit:
                case ExprKind::CharLit:
                case ExprKind::BoolLit:
                case ExprKind::NullLit:
                case ExprKind::Name:
                case ExprKind::GenericName:
                case ExprKind::Lambda:
                    return true;
                default:
                    return false;
            }
        }

        // An lvalue path: never bound to a value temporary (see the header). `Deref`
        // counts even when its pointer is a temporary: it still names a place the
        // emitter passes on as a reference.
        bool isPlace(const ExprPtr &e) {
            switch (e->kind) {
                case ExprKind::Name:
                case ExprKind::GenericName:
                case ExprKind::Deref:
                    return true;
                case ExprKind::Member:
                case ExprKind::Index:
                    return e->lhs && isPlace(e->lhs);
                default:
                    return false;
            }
        }

        // `&&`, `||` (and a future `?:`): operands that may not even be evaluated.
        bool isShortCircuit(const ExprPtr &e) {
            return e->kind == ExprKind::Binary && (e->text == "&&" || e->text == "||");
        }

        // Where an expression sits, which decides how much of it survives:
        // `Root` is the statement's own expression, `Value` is an operand, and
        // `Path` is a position that must stay a place (an assignment target, the
        // operand of `&`/`*`).
        enum class Slot { Root, Value, Path };

        class Flattener {
        public:
            List<StmtPtr> run(const List<StmtPtr> &stmts) {
                List<StmtPtr> out;
                walkStmts(stmts, out);
                return out;
            }

        private:
            // Per body, like the label counter in the linear pass.
            int next = 1;
            // The temporaries the statement being rewritten needs, in evaluation
            // order; they are emitted in front of it.
            List<StmtPtr> temps;

            Str freshTemp() { return Str("_sm_expr") + std::to_string(next++); }

            static ExprPtr nameExpr(const Str &name, const SourcePos &pos) {
                auto expr = std::make_shared<ast::Expr>();
                expr->kind = ExprKind::Name;
                expr->pos = pos;
                expr->text = name;
                return expr;
            }

            static StmtPtr tempDecl(const Str &name, const ExprPtr &init, const SourcePos &pos) {
                auto stmt = std::make_shared<Stmt>();
                stmt->kind = StmtKind::VarDecl;
                stmt->pos = pos;
                stmt->isVar = false;
                stmt->name = name;
                stmt->init = init;
                return stmt;
            }

            static StmtPtr blockStmt(List<StmtPtr> body, const SourcePos &pos) {
                auto stmt = std::make_shared<Stmt>();
                stmt->kind = StmtKind::Block;
                stmt->pos = pos;
                stmt->body = std::move(body);
                return stmt;
            }

            // Binds a value expression to a fresh temporary and returns its name.
            ExprPtr bind(const ExprPtr &e) {
                const Str name = freshTemp();
                temps.push_back(tempDecl(name, e, e->pos));
                return nameExpr(name, e->pos);
            }

            List<ExprPtr> flatList(const List<ExprPtr> &exprs) {
                List<ExprPtr> out;
                for (const ExprPtr &expr: exprs) out.push_back(flat(expr, Slot::Value));
                return out;
            }

            // A callee keeps its shape: `f`, `Res<T>.ok`, and a member call's
            // receiver path (`a[i].f(...)` stays a call on `a[i]`).
            ExprPtr flatCallee(const ExprPtr &e) {
                if (!e) return e;
                if (e->kind == ExprKind::Member) return flat(e, Slot::Path);
                if (e->kind == ExprKind::Name || e->kind == ExprKind::GenericName) return e;
                return flat(e, Slot::Value);
            }

            // A place stays a place; anything that produces a value is flattened as
            // a value (and bound if it is deeper than one operation).
            ExprPtr pathOrValue(const ExprPtr &e) {
                if (!e) return e;
                return isPlace(e) ? flat(e, Slot::Path) : flat(e, Slot::Value);
            }

            // The same node with its operands flattened. The copy is shallow and the
            // original is never touched.
            ExprPtr rebuild(const ExprPtr &e) {
                auto fresh = std::make_shared<ast::Expr>(*e);
                switch (e->kind) {
                    case ExprKind::Member:
                        fresh->lhs = pathOrValue(e->lhs);
                        break;
                    case ExprKind::Index:
                        fresh->lhs = pathOrValue(e->lhs);
                        fresh->rhs = flat(e->rhs, Slot::Value);
                        break;
                    case ExprKind::Call:
                        fresh->lhs = flatCallee(e->lhs);
                        fresh->args = flatList(e->args);
                        break;
                    case ExprKind::Binary:
                        fresh->lhs = flat(e->lhs, Slot::Value);
                        fresh->rhs = flat(e->rhs, Slot::Value);
                        break;
                    case ExprKind::Unary:
                    case ExprKind::Copy:
                        fresh->lhs = flat(e->lhs, Slot::Value);
                        break;
                    case ExprKind::Ref:
                    case ExprKind::Deref:
                        // `&x` boxes and `*x` borrows: both keep a place in place, or
                        // the address of a temporary would be taken.
                        fresh->lhs = pathOrValue(e->lhs);
                        break;
                    default:
                        break;
                }
                return fresh;
            }

            ExprPtr flat(const ExprPtr &e, Slot slot) {
                if (!e) return e;
                if (isSimple(e) || isShortCircuit(e)) return e;
                const ExprPtr built = rebuild(e);
                if (slot != Slot::Value) return built;
                if (isSimple(built) || isPlace(built)) return built;
                return bind(built);
            }

            // The statement plus the temporaries its expressions needed, scoped so a
            // jump can never cross one of them.
            StmtPtr withTemps(const StmtPtr &stmt) {
                if (temps.empty()) return stmt;
                temps.push_back(stmt);
                List<StmtPtr> body = std::move(temps);
                temps.clear();
                return blockStmt(std::move(body), stmt->pos);
            }

            void walkStmts(const List<StmtPtr> &stmts, List<StmtPtr> &out) {
                for (const StmtPtr &stmt: stmts) walkStmt(stmt, out);
            }

            void walkStmt(const StmtPtr &stmt, List<StmtPtr> &out) {
                if (!stmt) return;
                switch (stmt->kind) {
                    case StmtKind::Block: {
                        List<StmtPtr> inner;
                        walkStmts(stmt->body, inner);
                        out.push_back(blockStmt(std::move(inner), stmt->pos));
                        return;
                    }
                    case StmtKind::VarDecl: {
                        // The temporaries stay in this scope: the declared name is
                        // visible for the rest of the region. A region that declares
                        // anything is already a C++ block (`linear::lowerBody`),
                        // which is what keeps a jump from crossing them.
                        temps.clear();
                        auto fresh = std::make_shared<Stmt>(*stmt);
                        fresh->init = flat(stmt->init, Slot::Root);
                        for (const StmtPtr &temp: temps) out.push_back(temp);
                        out.push_back(fresh);
                        return;
                    }
                    case StmtKind::IfTrue:
                    case StmtKind::IfFalse: {
                        temps.clear();
                        const ExprPtr cond = flat(stmt->cond, Slot::Root);
                        auto fresh = std::make_shared<Stmt>(*stmt);
                        fresh->cond = cond;
                        out.push_back(withTemps(fresh));
                        return;
                    }
                    case StmtKind::Assign: {
                        temps.clear();
                        auto fresh = std::make_shared<Stmt>(*stmt);
                        fresh->target = flat(stmt->target, Slot::Path);
                        fresh->value = flat(stmt->value, Slot::Root);
                        out.push_back(withTemps(fresh));
                        return;
                    }
                    case StmtKind::Return: {
                        temps.clear();
                        auto fresh = std::make_shared<Stmt>(*stmt);
                        fresh->returnValue = flat(stmt->returnValue, Slot::Root);
                        out.push_back(withTemps(fresh));
                        return;
                    }
                    case StmtKind::ExprStmt: {
                        temps.clear();
                        auto fresh = std::make_shared<Stmt>(*stmt);
                        fresh->expr = flat(stmt->expr, Slot::Root);
                        out.push_back(withTemps(fresh));
                        return;
                    }
                    default:
                        // Label, Goto: no expressions. If/While/Switch/Break/Continue
                        // cannot appear here (linear::lowerBody removed them).
                        out.push_back(stmt);
                        return;
                }
            }
        };
    }

    List<StmtPtr> lowerExprs(const List<StmtPtr> &body) {
        Flattener flattener;
        return flattener.run(body);
    }
}

File: cppsrc/linear/ExpressionLowering.kt
// ExpressionLowering.kt
//
// Binds every expression deeper than one operation to a `_sm_expr<n>` temporary, one
// expression vocabulary for the emitter (impl_specs/linear-lowering.md, "Expression
// lowering"); runs after `linSimplifyBody`, counter per body. An lvalue path stays unbound;
// `&&`/`||`, whose operands are conditional, belong to the control-flow lowering.

package linear

// A literal, a name, a qualified name (`Res<T>`), or a lambda (lowered as its own body).
// The empty `None` node counts too: it is where a child that does not exist sits, and a
// missing child must not be bound to a temporary.
fun exprIsSimple(e: *AstXmlNode): Bool {
    val kind: AstNodeCategory = xmlKind(e)
    when (kind) {
        AstNodeCategory.None -> {
            return true
        }

        AstNodeCategory.ExprIntLit, AstNodeCategory.ExprFloatLit, AstNodeCategory.ExprStrLit,
        AstNodeCategory.ExprCharLit, AstNodeCategory.ExprBoolLit, AstNodeCategory.ExprNullLit,
        AstNodeCategory.ExprName, AstNodeCategory.ExprGenericName, AstNodeCategory.ExprLambda -> {
            return true
        }
    }
    return false
}

// An lvalue path, never bound to a value temporary; `ExprDeref` counts even when its
// pointer is a temporary.
fun exprIsPlace(e: *AstXmlNode): Bool {
    val kind: AstNodeCategory = xmlKind(e)
    when (kind) {
        AstNodeCategory.ExprName, AstNodeCategory.ExprGenericName, AstNodeCategory.ExprDeref -> {
            return true
        }

        AstNodeCategory.ExprMember, AstNodeCategory.ExprIndex -> {
            return exprIsPlace(xmlChildPtr(e, AstNodeKind.Receiver))
        }
    }
    return false
}

// Whether binding this expression to a temporary is *safe*: a borrow whose operand is not
// an lvalue (`*f()`) must stay inline, since `simse_addressOf`'s pointer only lasts for the
// call it is passed to (cppsrc/rtl/types.hpp).
fun exprIsBindable(e: *AstXmlNode): Bool {
    if (xmlKind(e) != AstNodeCategory.ExprDeref) {
        return true
    }
    return exprIsPlace(xmlChildPtr(e, AstNodeKind.Operand))
}

// `&&`, `||` (and a future `?:`): operands that may not even be evaluated.
fun exprIsShortCircuit(e: *AstXmlNode): Bool {
    if (xmlKind(e) != AstNodeCategory.ExprBinary) {
        return false
    }
    val op: Str = xmlAttr(e, AstNodeAttributeKind.Op)
    return op == "&&" || op == "||"
}

// A fresh node with the same role, kind and attributes - the role is what a parent looks a
// child up by (`xmlChild`/`xmlChildren`).
fun exprLike(like: *AstXmlNode, kids: *List<AstXmlNode>): AstXmlNode {
    return AstXmlNode(like.name, like.kind, copy(like.attributes), kids.toArray())
}

// The same node with every `role` child replaced, in order, by `replacements`; `like` is
// never modified.
fun exprReplaceRole(like: *AstXmlNode, role: AstNodeKind, replacements: *List<AstXmlNode>): AstXmlNode {
    var kids: List<AstXmlNode> = List<AstXmlNode>()
    val existing: List<AstXmlNode> = like.Children.toList()
    var seen: Int = 0
    for (*child in existing) {
        if (child.name == role) {
            if (seen < replacements.size()) {
                kids.append(replacements[seen])
            }
            seen = seen + 1
        } else {
            kids.append(child)
        }
    }
    while (seen < replacements.size()) {
        kids.append(replacements[seen])
        seen = seen + 1
    }
    return exprLike(like, kids)
}

// Where an expression sits. `Root` is the statement's own expression (a declaration's
// initializer, an assignment's value), `Value` is a position whose *value* is read, and
// `Path` is a position that must stay an alias (a call receiver, an assignment target, the
// operand of `&`/`*`).
enum class ExprSlot {
    Root,
    Value,
    Path
}

data class ExprFlattener(
    // Per body, like the label counter in the linear pass.
    var next: Int,
    // Set when an expression was bound to a temporary; a body already lowered comes back
    // false (`linLowerForEmission`).
    var changed: Bool,
    // Set when the statement being walked had something rebuilt under it; an untouched
    // statement is passed on as it is, so the pass does not rebuild every child
    // (`exprReplaceRole`).
    var touched: Bool
) {
    fun freshTemp(): Str {
        val id: Int = this.next
        this.next = this.next + 1
        return "_sm_expr" + id.toString()
    }

    fun roleName(role: AstNodeKind, name: *Str, line: Int, column: Int): AstXmlNode {
        return linName(role, name, line, column)
    }

    fun bind(e: *AstXmlNode, temps: *List<AstXmlNode>): AstXmlNode {
        val name: Str = this.freshTemp()
        temps.append(linSubjectDecl(name, e, xmlLine(e), xmlColumn(e)))
        this.changed = true
        return this.roleName(e.name, name, xmlLine(e), xmlColumn(e))
    }

    fun flatList(exprs: *List<AstXmlNode>, temps: *List<AstXmlNode>): List<AstXmlNode> {
        var out: List<AstXmlNode> = List<AstXmlNode>()
        for (*expr in exprs) {
            out.append(this.pathOrValue(expr, temps))
        }
        return out
    }

    // A callee keeps its shape: `f`, `Res<T>.ok`, and a member call's receiver path
    // (`a[i].f(...)` stays a call on `a[i]`).
    fun flatCallee(e: *AstXmlNode, temps: *List<AstXmlNode>): AstXmlNode {
        val kind: AstNodeCategory = xmlKind(e)
        when (kind) {
            AstNodeCategory.ExprMember -> {
                return this.flat(e, ExprSlot.Path, temps)
            }

            AstNodeCategory.ExprName, AstNodeCategory.ExprGenericName -> {
                return e
            }
        }
        return this.flat(e, ExprSlot.Value, temps)
    }

    fun pathOrValue(e: *AstXmlNode, temps: *List<AstXmlNode>): AstXmlNode {
        if (exprIsPlace(e)) {
            return this.flat(e, ExprSlot.Path, temps)
        }
        return this.flat(e, ExprSlot.Value, temps)
    }

    // The same node with its operands flattened. The original is never touched.
    fun rebuild(e: *AstXmlNode, temps: *List<AstXmlNode>): AstXmlNode {
        val kind: AstNodeCategory = xmlKind(e)
        this.touched = true
        when (kind) {
            AstNodeCategory.ExprMember -> {
                var receiver: List<AstXmlNode> = List<AstXmlNode>()
                receiver.append(this.pathOrValue(xmlChildPtr(e, AstNodeKind.Receiver), temps))
                return exprReplaceRole(e, AstNodeKind.Receiver, receiver)
            }

            AstNodeCategory.ExprIndex -> {
                var receiver: List<AstXmlNode> = List<AstXmlNode>()
                receiver.append(this.pathOrValue(xmlChildPtr(e, AstNodeKind.Receiver), temps))
                var index: List<AstXmlNode> = List<AstXmlNode>()
                index.append(this.flat(xmlChildPtr(e, AstNodeKind.Index), ExprSlot.Value, temps))
                val withReceiver: AstXmlNode = exprReplaceRole(e, AstNodeKind.Receiver, receiver)
                return exprReplaceRole(withReceiver, AstNodeKind.Index, index)
            }

            AstNodeCategory.ExprCall -> {
                var callee: List<AstXmlNode> = List<AstXmlNode>()
                callee.append(this.flatCallee(xmlChildPtr(e, AstNodeKind.Callee), temps))
                val argList: List<AstXmlNode> = xmlChildren(e, AstNodeKind.Arg)
                val args: List<AstXmlNode> = this.flatList(argList, temps)
                val withCallee: AstXmlNode = exprReplaceRole(e, AstNodeKind.Callee, callee)
                return exprReplaceRole(withCallee, AstNodeKind.Arg, args)
            }

            AstNodeCategory.ExprBinary -> {
                var lhs: List<AstXmlNode> = List<AstXmlNode>()
                lhs.append(this.flat(xmlChildPtr(e, AstNodeKind.Lhs), ExprSlot.Value, temps))
                var rhs: List<AstXmlNode> = List<AstXmlNode>()
                rhs.append(this.flat(xmlChildPtr(e, AstNodeKind.Rhs), ExprSlot.Value, temps))
                val withLhs: AstXmlNode = exprReplaceRole(e, AstNodeKind.Lhs, lhs)
                return exprReplaceRole(withLhs, AstNodeKind.Rhs, rhs)
            }

            AstNodeCategory.ExprUnary, AstNodeCategory.ExprCopy -> {
                var operand: List<AstXmlNode> = List<AstXmlNode>()
                operand.append(this.flat(xmlChildPtr(e, AstNodeKind.Operand), ExprSlot.Value, temps))
                return exprReplaceRole(e, AstNodeKind.Operand, operand)
            }

            AstNodeCategory.ExprRef, AstNodeCategory.ExprDeref -> {
                // `&x` boxes and `*x` borrows: the operand stays a place, or the address
                // of a temporary would be taken.
                var operand: List<AstXmlNode> = List<AstXmlNode>()
                operand.append(this.pathOrValue(xmlChildPtr(e, AstNodeKind.Operand), temps))
                return exprReplaceRole(e, AstNodeKind.Operand, operand)
            }
        }
        return e
    }

    fun flat(e: *AstXmlNode, slot: ExprSlot, temps: *List<AstXmlNode>): AstXmlNode {
        if (exprIsSimple(e) || exprIsShortCircuit(e)) {
            return e
        }
        val built: AstXmlNode = this.rebuild(e, temps)
        if (slot != ExprSlot.Value) {
            return built
        }
        // A *value* position is one operation deep: `self.x` becomes its own temporary,
        // so only the alias positions above keep a path unbound.
        if (exprIsSimple(built) || !exprIsBindable(built)) {
            return built
        }
        return this.bind(built, temps)
    }

    // The statement plus the temporaries its expressions needed, scoped so a jump can never
    // cross one of them.
    fun withTemps(stmt: AstXmlNode, temps: *List<AstXmlNode>): AstXmlNode {
        if (temps.size() == 0) {
            return stmt
        }
        temps.append(stmt)
        return linBlock(temps, xmlLine(stmt), xmlColumn(stmt))
    }

    // Whether any statement had a rebuild under it: a child's walk clears `touched` as it
    // goes, so the answer is the disjunction.
    fun walkStmts(stmts: *List<AstXmlNode>, out: *List<AstXmlNode>): Bool {
        var any: Bool = false
        for (*stmt in stmts) {
            this.walkStmt(stmt, out)
            if (this.touched) {
                any = true
            }
        }
        return any
    }

    fun walkStmt(stmt: *AstXmlNode, out: *List<AstXmlNode>): Unit {
        val kind: AstNodeCategory = xmlKind(stmt)
        var temps: List<AstXmlNode> = List<AstXmlNode>()
        this.touched = false
        when (kind) {
            AstNodeCategory.StmtBlock -> {
                var inner: List<AstXmlNode> = List<AstXmlNode>()
                val anyInner: Bool =
                    this.walkStmts(xmlChildren(xmlChildPtr(stmt, AstNodeKind.Body), AstNodeKind.Stmt), inner)
                this.touched = anyInner
                if (!anyInner) {
                    out.append(stmt)
                    return
                }
                out.append(linBlock(inner, xmlLine(stmt), xmlColumn(stmt)))
                return
            }

            AstNodeCategory.StmtVarDecl -> {
                // The temporaries stay in this scope (the declared name is visible for the
                // rest of the region, which is already a C++ block, so no jump crosses them).
                var init: List<AstXmlNode> = List<AstXmlNode>()
                init.append(this.flat(xmlChildPtr(stmt, AstNodeKind.Init), ExprSlot.Root, temps))
                if (!this.touched) {
                    out.append(stmt)
                    return
                }
                var i: Int = 0
                while (i < temps.size()) {
                    out.append(temps[i])
                    i = i + 1
                }
                out.append(exprReplaceRole(stmt, AstNodeKind.Init, init))
                return
            }

            AstNodeCategory.StmtIfTrue, AstNodeCategory.StmtIfFalse -> {
                // The condition is a value position: a non-name becomes its own temporary,
                // so the jump is all the statement does.
                var cond: List<AstXmlNode> = List<AstXmlNode>()
                cond.append(this.flat(xmlChildPtr(stmt, AstNodeKind.Cond), ExprSlot.Value, temps))
                if (!this.touched) {
                    out.append(stmt)
                    return
                }
                out.append(this.withTemps(exprReplaceRole(stmt, AstNodeKind.Cond, cond), temps))
                return
            }

            AstNodeCategory.StmtAssign -> {
                var target: List<AstXmlNode> = List<AstXmlNode>()
                target.append(this.flat(xmlChildPtr(stmt, AstNodeKind.Target), ExprSlot.Path, temps))
                var value: List<AstXmlNode> = List<AstXmlNode>()
                value.append(this.flat(xmlChildPtr(stmt, AstNodeKind.Value), ExprSlot.Root, temps))
                if (!this.touched) {
                    out.append(stmt)
                    return
                }
                val withTarget: AstXmlNode = exprReplaceRole(stmt, AstNodeKind.Target, target)
                out.append(this.withTemps(exprReplaceRole(withTarget, AstNodeKind.Value, value), temps))
                return
            }

            AstNodeCategory.StmtReturn -> {
                var value: List<AstXmlNode> = List<AstXmlNode>()
                value.append(this.flat(xmlChildPtr(stmt, AstNodeKind.Value), ExprSlot.Value, temps))
                if (!this.touched) {
                    out.append(stmt)
                    return
                }
                out.append(this.withTemps(exprReplaceRole(stmt, AstNodeKind.Value, value), temps))
                return
            }

            AstNodeCategory.StmtExprStmt -> {
                var expr: List<AstXmlNode> = List<AstXmlNode>()
                expr.append(this.flat(xmlChildPtr(stmt, AstNodeKind.Expr), ExprSlot.Root, temps))
                if (!this.touched) {
                    out.append(stmt)
                    return
                }
                out.append(this.withTemps(exprReplaceRole(stmt, AstNodeKind.Expr, expr), temps))
                return
            }

            AstNodeCategory.StmtYield -> {
                // `yield e` hands a value out, so its expression is a value position like a
                // `return`'s: one operand, or a temporary. (The state-machine pass replaces
                // the statement afterwards.)
                var value: List<AstXmlNode> = List<AstXmlNode>()
                value.append(this.flat(xmlChildPtr(stmt, AstNodeKind.Value), ExprSlot.Value, temps))
                if (!this.touched) {
                    out.append(stmt)
                    return
                }
                out.append(this.withTemps(exprReplaceRole(stmt, AstNodeKind.Value, value), temps))
                return
            }
        }
        // Label/Goto hold no expressions; If/While/Break/Continue are gone by now
        // (`linLowerBody`).
        out.append(stmt)
    }
}

// Lowers the expressions of one linear body; the counter restarts per body.
fun linLowerExprs(body: *List<AstXmlNode>): LinLowered {
    var flattener: ExprFlattener = ExprFlattener(1, false, false)
    var out: List<AstXmlNode> = List<AstXmlNode>()
    flattener.walkStmts(body, out)
    return LinLowered(out, flattener.changed)
}
